### ISSUE TITLE
[paradoxalarm] Add access to the different Alarm States (Armed, Armed in Stay, Was Is in Alarm)

### ISSUE THREAD

**Issue by Mark-VG** [openhab/openhab-addons#14608]
The current version of the Binding seems to groups the different Armed states into the general Armed state, so it is not possible to see what type of Arm in the partition is in.

Armed - is usually used to arm all zones and exit the partition/area
Armed Stay - allows arming selected Zones so that you can "Stay" in the area, but certain selected Zones are "bypassed/ignored". This allows movement in select areas.

It would be really nice to have access to the additional states, either as distinct "switches" such as:

Armed Exit - On/Off
Armed Stay - On/Off
Armed Instant - On/Off
Was Is in Alarm - On/Off

The other option would be to Just Have a String in the Current State String to denote the current state, but as you mentioned that would require:

```
The other issue is that in any rules that we have we need to have "if (armed || armed in stay || armed in other way) do ...."
```
I think the separate Channel option would be easier to work with, though both have there pros and cons.

The Caddx binding which I used to use had separate channels for each Armed state

**Comment by theater** [comment]
Hi Mark,
we could implement a consolidated status but we need to define the priority (i.e. if multiple states are present, which one to take into account). Have you thought about this because I can implement something but it may not be the right one. :)

For example: If we have "was in alarm", "armed in stay" and "armed", which one to take into consideration?
Currently the main state of the partition is calculated like this - "In alarm" is with highest priority, then all the "armed" states are treated as "Armed", all the others consider "disarmed". 

See the coding:
` public String getMainState() {
        if (isInAlarm) {
            return "InAlarm";
        } else {
            return isArmed || isArmedInAway || isArmedInStay || isArmedInNoEntry ? "Armed" : "Disarmed";
        }
    }`

Cheers,
Konstantin

**Comment by Mark-VG** [comment]
Hi Konstantin

If you did a consolidated status, would you be able to distinguish between the different status and return the correct string?

From what I know you cannot be ArmedInStay as well as ArmedInAway or ArmedInNoEntry at the same time. But you might get be able to be Armed?

Maybe (not my area)???

```
public String getMainState() {
if (isInAlarm) {
return "InAlarm";
} else {
if (isArmedInAway) {
return "Armed Away";
} else {
if (isArmedInStay) {
return "Armed Stay";
} else {
if (isArmedInNoEntry ) {
return "Armed No Entry";
}
}
```
But a more flexible option might be to Have the Overall Status Being In Alarm or Armed or Disarmed - could even add the extra one) with a more detailed channel for each state being

InAlarm On/Off
ArmedAway On/Off
ArmedStay On/Off
ArmedNoEntry On/Off

Then you can choose which works better for the scenario?

**Comment by theater** [comment]
Yes. 
These are the two feasible options. 
If if these statuses are exclusive (i.e. if you can be either Armed or Armed in stay or Armed in no entry), maybe it's better to put it in the overall state because we will have the most exact state in the overall state. 
I think I'm leaning towards this solution because otherwise you have to assign the channels to a separate items and then check each item's status, if you want to use them in some rules for example. (or maybe you can check thing's channels... I'm not sure about this).

**Comment by theater** [comment]
OK... in order to still keep this backwards compatible, I think what we can do is the following:
in case of isInAlarm or isInSilentAlarm or isInAudibleAlarm or isInFireAlarm => main state = In Alarm
in case of isArmed or isArmedInAway or isArmedInStay or isArmedInNoEntry => main state = Armed / Disarmed (if none of these is set)

Implement additional channel (only one) which returns calculated sub-state (or "exact state") of all these which is the exact state, i.e. one of these 8 or "disarmed" if none of these 8 bits is set. 
With this approach whoever has rules based on the main state will still be able to use them, whoever needs exact states, can use the sub-state (or exact state) channel.

What do you think?

**Comment by theater** [comment]
You can try the new channel using this release: https://github.com/theater/openhab-addons/releases/tag/v4.0.0-alpha6

**Comment by Mark-VG** [comment]
I am getting Armed and Disarmed only. When Armed Stay I just get Armed.

**Comment by theater** [comment]
Can you raise log level to trace and send me a random log from a retrieve
during arm with stay?

**Comment by Mark-VG** [comment]
Hi. Hope this is enough. Partition 4 is Armed STAY
[paradox.log](https://github.com/openhab/openhab-addons/files/10983273/paradox.log)

**Comment by theater** [comment]
Hi, I've uploaded a new version in the same release. No fix but more logging. I need to see what happens in the parsing. Hopefully now we will have the full state like in the zones for better debugging. Sorry for the inconvenience but no time to test locally...

**Comment by Mark-VG** [comment]
Thank you... I have installed an running. Logs as follows:
[paradox1.log](https://github.com/openhab/openhab-addons/files/10993956/paradox1.log)

I did notice the build seems to be the same as yesterday? Is that correct?
```
260 │ Active │  80 │ 4.0.0.202303151448     │ openHAB Add-ons :: Bundles :: ParadoxAlarm Binding
263 │ Active │  80 │ 4.0.0.202303151448     │ openHAB Add-ons :: Bundles :: ParadoxAlarm Binding
```

**Comment by theater** [comment]
OK. Added also a new table with the possible detailed state values in the README.

**Comment by Mark-VG** [comment]
Hi

That looks really great. Have done a basic test and reports the Armed Stay well. Have removed and readded all the Things and shows up great

One small comment regarding the default naming:
![image](https://user-images.githubusercontent.com/84638697/225875786-a08be329-0241-4e46-a020-917beb56c302.png)

Just the Case of the partition in the name:
```
Partition_House_Detailed_partition_State
```

In the other examples all the words are Upper Case for each Word. But not this one. Not sure if you want to make it the same?

I could not find the modified README though.

Thanks again.

**Comment by Mark-VG** [comment]
All these additions are working really well. Thank you so much.

Do you think there is still an opportunity to make a cosmetic change since you have submitted the PR already?

To match the Keypad display for the details States, could the String be changed? I know these will not be changed once released as they would affect anyone who is using them.

![image](https://user-images.githubusercontent.com/84638697/225960540-e529d4c4-6039-4614-9caf-c00ffd17d502.png)


Maybe to 

Away Armed
Stay Armed
NoEntry Armed

and Also Possibly :
Silent Alarm
Audible Alarm
Fire Alarm

I am trying to make my widget reflect the Keypad as best I can - not always possible:


![Statuses](https://user-images.githubusercontent.com/84638697/225961334-75bb3489-605a-4951-9ff7-0770600bb9c6.gif)

Just something you could consider - most certainly not the end of the world. Couldn't be happier with what you have managed to add already.

**Comment by theater** [comment]
Yeah. No worries. This change is trivial and totally makes sense.
Will do that today and will inform you so you could have a look.

**Comment by theater** [comment]
https://github.com/theater/openhab-addons/releases/tag/v4.0.0-beta1

**Comment by Mark-VG** [comment]
Thank you. Looks good. Will continue to work with the latest one and let you know. 

Did you decide to leave the default Linked Item Name as is?

```
Partition_Beams_Detailed_partition_State
```
vs
```
Partition_Beams_Detailed_**P**artition_State
```
Just asking because the the Widget I am working on will relay on the names of all the Channel Items, so trying to match the defaults as far as I can, and this one is a bit different to the others because of the lowercase p for Partiton.

Hope you have a good week-end.

**Comment by theater** [comment]
Wooah... I don't know where this default value comes from. Probably the UI
is building it based on some reflection of the field names in the class...
I changed the typeID in the metadata file to be all capitalized and I hoped
it's coming from there. Can you try it again with a new item/channel?
See below:
<channel-type id="detailedState">
<item-type>String</item-type>
<label>*Detailed Partition State*</label>
<description>The detailed state of partition (contains sub-states like Stay
in Armed, Armed no entry, etc)</description>
<state readOnly="true" pattern="%s"/>
</channel-type>

**Comment by Mark-VG** [comment]
I have deleted and re-added the Partition THING, and now the default is all CAPS for first letters:

![image](https://user-images.githubusercontent.com/84638697/226105064-2cd94e99-49c7-43f6-9516-5b7458f20ab6.png)

I have noticed that most of the linked ITEM defaults have also changed to have _ between each word. I have just recreted to match that.

Maybe 4.0.0 does something different.

It is a pity that openHAB requires a delete and re-add to pick these changes.

Thank yous o much

**Comment by theater** [comment]
Well it's not a big deal as you usually do it only once initially and you
can enter there anything.
OK, so for now all requirements seem to be settled. Let's hope someone
finds time to review both PRs soon so we can merge them in the main branch.

Cheers,
K.

**Comment by Mark-VG** [comment]
Yes. All great. Hopefully reviewed and merged soon.
